


# 🎓 Projeto de Conclusão: Sistema Inteligente de Análise de Perfil Financeiro
## Tema 4: Decisão Inteligente (Previsão de Churn - Telco)

## ▶ 1. Introdução

O setor de Telecomunicações é um dos pilares da economia digital, caracterizado por uma alta competitividade e serviços com baixo diferencial perceptível (commodity). Nesse cenário, a rotatividade de clientes, conhecida como **Churn Rate**, representa um dos maiores desafios estratégicos. A perda de um cliente não significa apenas a interrupção de uma receita mensal, mas também o desperdício do investimento realizado em sua aquisição (**CAC - Customer Acquisition Cost**).

Estudos indicam que manter um cliente atual pode ser de **5 a 25 vezes mais barato** do que adquirir um novo ([Harvard Businness Review - The Value of Keeping New Customers](https://hbr.org/2014/10/the-value-of-keeping-the-right-customers)). Portanto, a capacidade de prever o comportamento de abandono antes que ele ocorra permite que a empresa transicione de uma postura reativa para uma estratégia proativa. Ao utilizar dados para antecipar o risco de saída, a organização pode personalizar ofertas, melhorar o suporte técnico e ajustar planos financeiros, maximizando o **LTV (Lifetime Value)** e garantindo a sustentabilidade do negócio a longo prazo.

### 📝 Nota sobre a Metodologia e Dados

Este projeto utiliza uma abordagem analítica baseada em evidências reais do mercado de telecomunicações. Embora o conjunto de dados utilizado ([Telco Customer Churn](https://www.kaggle.com/datasets/blastchar/telco-customer-churn)) seja uma base de dados aberta disponibilizada pela IBM para fins de estudo e desenvolvimento de modelos, os padrões e comportamentos aqui identificados refletem desafios reais enfrentados por empresas do setor. O foco desta análise não é apenas a precisão estatística, mas a demonstração de como uma arquitetura de dados bem estruturada pode transformar registros brutos em decisões estratégicas.




## 🎯 2. Objetivo
O objetivo central deste projeto é desenvolver uma solução de **Decision Intelligence** capaz de prever o risco de cancelamento (churn) de clientes em uma empresa de telecomunicações, fornecendo subsídios estratégicos para ações de retenção personalizadas.

### 2.1 Objetivos de Negócio

* **Redução de Churn:** Identificar precocemente clientes com alta probabilidade de abandono para reduzir a perda de receita mensal.
* **Apoio à Decisão:** Criar simulações de cenários e recomendações baseadas em dados para os times de Marketing e Customer Success.
* **Otimização de ROI:** Focar esforços de retenção nos clientes de maior valor monetário e maior risco.

### 2.2 Objetivos Técnicos
* **Engenharia de Dados:** Realizar a ingestão, limpeza e tratamento de dados brutos, garantindo a integridade estatística das variáveis financeiras e de consumo.

* **Análise Exploratória (EDA):** Identificar correlações e padrões de comportamento que segmentam clientes retidos de clientes perdidos.

* **Modelagem Preditiva:** Treinar, avaliar e validar um modelo de Machine Learning para classificação binária de churn.

* **Visualização Analítica:** Desenvolver um dashboard interativo no Power BI que conecte as predições do modelo com indicadores de desempenho (KPIs) executivos.

## 📖 3. Dicionário de Dados (Versão Traduzida)

Abaixo, descrevemos as variáveis que serão utilizadas ao longo da análise, após o processo de tradução e adequação para o contexto da consultoria:

| Variável | Descrição | Tipo |
| :--- | :--- | :--- |
| **id_cliente** | Identificador único de cada cliente. | Categórico/ID |
| **genero** | Gênero do cliente (Masculino/Feminino). | Categórico |
| **idoso** | Indica se o cliente tem 65 anos ou mais (1: Sim, 0: Não). | Numérico/Binário |
| **possui_parceiro** | Indica se o cliente possui um(a) parceiro(a) (Sim/Não). | Categórico |
| **possui_dependentes** | Indica se o cliente possui dependentes (Sim/Não). | Categórico |
| **tempo_contrato_meses**| Número de meses que o cliente permaneceu na empresa. | Numérico |
| **servico_telefone** | Indica se o cliente possui serviço de telefonia (Sim/Não). | Categórico |
| **multiplas_linhas** | Se o cliente tem múltiplas linhas (Sim, Não, Sem serviço). | Categórico |
| **servico_internet** | Provedor de internet do cliente (DSL, Fibra óptica, Não). | Categórico |
| **seguranca_online** | Se possui serviço de segurança online (Sim/Não/Sem internet). | Categórico |
| **backup_online** | Se possui serviço de backup online (Sim/Não/Sem internet). | Categórico |
| **protecao_dispositivo**| Se possui plano de proteção de dispositivo (Sim/Não/Sem internet).| Categórico |
| **suporte_tecnico** | Se possui suporte técnico prioritário (Sim/Não/Sem internet).| Categórico |
| **streaming_tv** | Se o cliente possui streaming de TV (Sim/Não/Sem internet). | Categórico |
| **streaming_filmes** | Se o cliente possui streaming de filmes (Sim/Não/Sem internet).| Categórico |
| **tipo_contrato** | Tipo de contrato (Mensal, Um ano, Dois anos). | Categórico |
| **fatura_digital** | Se o cliente optou por fatura sem papel (Sim/Não). | Categórico |
| **metodo_pagamento** | Método de pagamento (Cheque eletrônico, Boleto, Cartão, etc). | Categórico |
| **valor_mensal** | Valor cobrado do cliente mensalmente. | Numérico |
| **valor_total** | O valor total cobrado do cliente acumulado. | Numérico |
| **cancelamento** | **Variável Alvo:** Indica se o cliente cancelou o serviço (Sim/Não).| Categórico |

## ⚙️ 4. Setup

In [3]:

# Importação de bibliotecas
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import sqlite3
import os
import warnings

# Configurações de visualização
%matplotlib inline
sns.set_theme(style="whitegrid", palette="viridis")
plt.rcParams['figure.figsize'] = (10, 6)

# Configurações de exibição do Pandas
pd.set_option('display.max_columns', None)

# Ignorar warnings
warnings.simplefilter(action="ignore")



In [4]:
#Conexão com o repositório do github

nome_repositorio = 'hackathon-elas-tech-grupo7'
url_repositorio = 'https://github.com/joyexplorer/hackathon-elas-tech-grupo7.git'

if not os.path.exists(nome_repositorio):
  print(f"Clonando o repositório {nome_repositorio}...")
  !git clone {url_repositorio}
else:
  print(f"O repositório {nome_repositorio} já existe. Atualizando com as últimas mudanças...")
  %cd {nome_repositorio}
  !git pull
  %cd ..

os.chdir(f'/content/{nome_repositorio}')
print(f"Diretório atual: {os.getcwd()}")

Clonando o repositório hackathon-elas-tech-grupo7...
Cloning into 'hackathon-elas-tech-grupo7'...
remote: Enumerating objects: 134, done.
remote: Counting objects: 100% (134/134), done.
remote: Compressing objects: 100% (113/113), done.
remote: Total 134 (delta 60), reused 4 (delta 3), pack-reused 0 (from 0)
Receiving objects: 100% (134/134), 583.52 KiB | 3.67 MiB/s, done.
Resolving deltas: 100% (60/60), done.
Diretório atual: /content/hackathon-elas-tech-grupo7


## 📥 5. Ingestão de Dados e Ajustes Iniciais

### 📦 Origem do Dataset
Os dados foram extraídos do repositório de amostras da IBM Guide e disponibilizados via Kaggle. O dataset contém informações de 7.043 clientes de uma empresa fictícia de telecomunicações na Califórnia no terceiro trimestre e apresenta desafios comuns de engenharia de dados, como tipos de variáveis incorretos e dados faltantes mascarados.

In [16]:
df = pd.read_csv('/content/hackathon-elas-tech-grupo7/dados/brutos/telco-customer-churn.csv')

In [17]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


## 🔍 5.1. Conhecendo o Dataset
O primeiro passo após ingerir o dataset é *conhecê-lo*: explorar sua estrutura geral, como número de linhas e colunas, nomes das features, tipo e conteúdo dos dados. Essa etapa é fundamental para que a próxima etapa de limpeza e tratamento de dados seja completa e adequada neste contexto.

In [19]:
# São 21 colunas
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

### 🔤 Tradução das colunas
Embora o dataset original utilize a nomenclatura em inglês, optamos pela tradução integral de todas as variáveis para o português. Esta decisão reflete o contexto estratégico deste projeto: simulamos aqui a entrega de uma **consultoria de dados** especializada para uma empresa brasileira.

O objetivo é garantir que este notebook funcione como uma apresentação executiva para o Board de Diretores. Ao eliminarmos a barreira do idioma, facilitamos a democratização dos dados (data democratization) e permitimos que os stakeholders foquem no que realmente importa: a compreensão dos insights gerados, a lógica da análise preditiva e o valor estratégico que este tipo de serviço oferece para a retenção de clientes e saúde financeira da companhia no Brasil.



In [22]:
#Opção de renomear as colunas para português / Só se acharem necessário
#TotalCharges está como objeto e deveria ser numérico
#Talvez não precisamos de todas as colunas

df.rename(columns={
    'customerID': 'id_cliente',
    'gender': 'genero',
    'SeniorCitizen': 'idoso',
    'Partner': 'possui_parceiro',
    'Dependents': 'possui_dependentes',
    'tenure': 'tempo_contrato_meses',
    'PhoneService': 'servico_telefone',
    'MultipleLines': 'multiplas_linhas',
    'InternetService': 'servico_internet',
    'OnlineSecurity': 'seguranca_online',
    'OnlineBackup': 'backup_online',
    'DeviceProtection': 'protecao_dispositivo',
    'TechSupport': 'suporte_tecnico',
    'StreamingTV': 'streaming_tv',
    'StreamingMovies': 'streaming_filmes',
    'Contract': 'tipo_contrato',
    'PaperlessBilling': 'fatura_digital',
    'PaymentMethod': 'metodo_pagamento',
    'MonthlyCharges': 'valor_mensal',
    'TotalCharges': 'valor_total',
    'Churn': 'cancelamento'
}, inplace=True)



### 🔬 Explorando o conteúdo

In [18]:
#Quantidade de linhas e colunas dos nossos dados
df.shape

(7043, 21)

In [23]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id_cliente            7043 non-null   object 
 1   genero                7043 non-null   object 
 2   idoso                 7043 non-null   int64  
 3   possui_parceiro       7043 non-null   object 
 4   possui_dependentes    7043 non-null   object 
 5   tempo_contrato_meses  7043 non-null   int64  
 6   servico_telefone      7043 non-null   object 
 7   multiplas_linhas      7043 non-null   object 
 8   servico_internet      7043 non-null   object 
 9   seguranca_online      7043 non-null   object 
 10  backup_online         7043 non-null   object 
 11  protecao_dispositivo  7043 non-null   object 
 12  suporte_tecnico       7043 non-null   object 
 13  streaming_tv          7043 non-null   object 
 14  streaming_filmes      7043 non-null   object 
 15  tipo_contrato        

#### 🧐 Diagnóstico da Estrutura de Dados
Após a inspeção técnica das 21 variáveis e 7.043 registros, identificamos pontos críticos que guiarão nossa próxima etapa de limpeza:

**Inconsistência de Tipagem:** A variável valor_total (originalmente TotalCharges) está classificada como object (texto). Por se tratar de uma métrica financeira fundamental para o cálculo de faturamento e churn, ela precisa ser convertida para float64 para permitir análises estatísticas.

**Variáveis Categóricas Dominantes:** Notamos que a grande maioria dos dados (18 colunas) é do tipo categórico (object). Isso indica que precisaremos realizar técnicas de encoding (transformação em números) antes de alimentar o modelo de Machine Learning.

**Falsos Positivos de Integridade:** Embora o resumo indique que não há valores nulos (7043 non-null), é possível que neste dataset existam valores ausentes estejam mascarados como espaços em branco (" "), especialmente na coluna de valores totais. Investigaremos essa hipótese na fase de tratamento.

**Métricas Numéricas Diretas:** Possuímos apenas 3 colunas numéricas nativas iniciais: idoso (binária), tempo_contrato_meses e valor_mensal.

In [24]:
# Verificando a quantidade de valores únicos por coluna
df.nunique()

,0
id_cliente,7043
genero,2
idoso,2
possui_parceiro,2
possui_dependentes,2
tempo_contrato_meses,73
servico_telefone,2
multiplas_linhas,3
servico_internet,3
seguranca_online,3


#SQL

In [ ]:
# =====================================================
# CONFIGURAÇÃO INICIAL: Conexão ao banco SQLite
# importar bibliotecas e criar a conexão principal (conn)
# com o banco de dados para execução das queries SQL.
# =====================================================
conn = sqlite3.connect('/content/hackathon-elas-tech-grupo7/sql/telco_churn_hackathon.db')

In [ ]:
query = "SELECT * FROM clientes LIMIT 5;"
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [ ]:
# =====================================================
# VERIFICAÇÃO INICIAL:
# Lista todas as tabelas disponíveis no banco de dados.
# e validar o nome exato da tabela antes das análises.
# =====================================================

query = """
SELECT name FROM sqlite_master WHERE type='table';
"""

pd.read_sql_query(query, conn)

,name
0,clientes


In [ ]:
# =====================================================
# KPI 1: Total de clientes da base
# Esta consulta contabiliza o número total de registros
# (clientes únicos) presentes na base de dados.
# Serve como referência principal para cálculos
# percentuais futuros, como churn rate.
# =====================================================

query = """
SELECT COUNT(*) AS total_clientes
FROM clientes;
"""

total_clientes = pd.read_sql_query(query, conn)
total_clientes

,total_clientes
0,7032


### Aqui vemos que cerca de 1 em cada 4 clientes cancela, uma alta taxa de churn


In [ ]:
# =====================================================
# KPI 2: Taxa geral de churn
# Calcula:
# - Número de clientes que cancelaram
# - Número de clientes ativos
# - Percentual de cada grupo
# mede a taxa de evasão da empresa.
# =====================================================

query = """
SELECT
    Churn,
    COUNT(*) AS total_clientes,
    ROUND(COUNT(*) * 100.0 / (SELECT COUNT(*) FROM clientes), 2) AS percentual
FROM clientes
GROUP BY Churn;
"""

taxa_churn = pd.read_sql_query(query, conn)
taxa_churn

,Churn,total_clientes,percentual
0,No,5163,73.42
1,Yes,1869,26.58


### Base possui alto valor financeiro, $455.661/mês

In [ ]:
# =====================================================
# KPI 3: Receita mensal total
# Soma todos os valores mensais pagos pelos clientes,
# representando o faturamento recorrente mensal atual.
# =====================================================

query = """
SELECT
    ROUND(SUM(MonthlyCharges), 2) AS receita_mensal_total
FROM clientes;
"""

receita_total = pd.read_sql_query(query, conn)
receita_total

,receita_mensal_total
0,455661.0


### Receita associada aos clientes perdidos, aproximadamente 30,5% da receita mensal total está em risco, Resultado: $139.130,85/mês

In [ ]:
# =====================================================
# KPI 4: Receita mensal em risco
# Soma o valor mensal associado aos clientes
# que já cancelaram o serviço.

# Essa métrica ajuda a estimar impacto financeiro
# da evasão de clientes.
# =====================================================

query = """
SELECT
    ROUND(SUM(MonthlyCharges), 2) AS receita_em_risco
FROM clientes
WHERE Churn = 'Yes';
"""

receita_risco = pd.read_sql_query(query, conn)
receita_risco

,receita_em_risco
0,139130.85


### Clientes churnados pagam mais, A empresa não está perdendo apenas clientes — está perdendo clientes de maior valor (“Clientes mais lucrativos estão cancelando.”)

In [ ]:
# =====================================================
# KPI 5: Ticket médio por churn
# Calcula o gasto médio mensal de clientes
# separados entre churn e não churn.
#
# Permite avaliar se clientes perdidos
# possuem maior valor monetário.
# =====================================================

query = """
SELECT
    Churn,
    ROUND(AVG(MonthlyCharges), 2) AS ticket_medio
FROM clientes
GROUP BY Churn;
"""

ticket_medio = pd.read_sql_query(query, conn)
ticket_medio

,Churn,ticket_medio
0,No,61.31
1,Yes,74.44


### Contrato mensal concentra maior churn, Month-to-month: ~42,7% || Two year: ~2,8%, ou seja, Fidelização contratual reduz drasticamente churn

In [ ]:
# =====================================================
# KPI 6: Churn por tipo de contrato
# Analisa distribuição de churn conforme:
# - Contrato mensal
# - Contrato anual
# - Contrato bienal
#
# Identifica contratos mais vulneráveis.
# =====================================================

query = """
SELECT
    Contract,
    Churn,
    COUNT(*) AS total
FROM clientes
GROUP BY Contract, Churn
ORDER BY Contract;
"""

churn_contrato = pd.read_sql_query(query, conn)
churn_contrato

,Contract,Churn,total
0,Month-to-month,No,2220
1,Month-to-month,Yes,1655
2,One year,No,1306
3,One year,Yes,166
4,Two year,No,1637
5,Two year,Yes,48


In [ ]:
tabela_comparativa = churn_contrato.pivot(index='Contract', columns='Churn', values='total') #Melhor visual da tabela
tabela_comparativa

Churn,No,Yes
Contract,,
Month-to-month,2220,1655
One year,1306,166
Two year,1637,48


### Métodos automáticos apresentam churn muito menor

In [ ]:
# =====================================================
# KPI 7: Churn por método de pagamento
# Avalia se certos métodos de pagamento
# possuem maior taxa de cancelamento.
#
# Pode revelar padrões financeiros
# e oportunidades de retenção.
# =====================================================

query = """
SELECT
    PaymentMethod,
    Churn,
    COUNT(*) AS total
FROM clientes
GROUP BY PaymentMethod, Churn;
"""

churn_pagamento = pd.read_sql_query(query, conn)
churn_pagamento

,PaymentMethod,Churn,total
0,Bank transfer (automatic),No,1284
1,Bank transfer (automatic),Yes,258
2,Credit card (automatic),No,1289
3,Credit card (automatic),Yes,232
4,Electronic check,No,1294
5,Electronic check,Yes,1071
6,Mailed check,No,1296
7,Mailed check,Yes,308


### Clientes que cancelam permanecem menos da metade do tempo.

In [ ]:
# =====================================================
# KPI 8: Tempo médio de permanência
# Mede quantos meses, em média,
# clientes churnados e não churnados
# permanecem na empresa.
#
# Fundamental para entender
# o ciclo de vida do cliente.
# =====================================================

query = """
SELECT
    Churn,
    ROUND(AVG(tenure), 2) AS media_tempo_contrato
FROM clientes
GROUP BY Churn;
"""

tempo_permanencia = pd.read_sql_query(query, conn)
tempo_permanencia

,Churn,media_tempo_contrato
0,No,37.65
1,Yes,17.98


### Há clientes antigos e valiosos cancelando, Padrões: MonthlyCharges > $110, Tenure alto, TotalCharges elevado

In [ ]:
# =====================================================
# KPI 9: Clientes de maior valor em risco
# Lista clientes churnados com maiores cobranças mensais.
#
# Ajuda a identificar clientes prioritários
# para estratégias de retenção premium.
# =====================================================

query = """
SELECT
    customerID,
    MonthlyCharges,
    TotalCharges,
    tenure
FROM clientes
WHERE Churn = 'Yes'
ORDER BY MonthlyCharges DESC
LIMIT 20;
"""

clientes_risco = pd.read_sql_query(query, conn)
clientes_risco

,customerID,MonthlyCharges,TotalCharges,tenure
0,8199-ZLLSA,118.35,7804.15,67
1,2889-FPWRM,117.80,8684.80,72
2,2302-ANTDP,117.45,5438.90,48
3,9053-JZFKV,116.20,7752.30,67
4,1444-VVSGW,115.65,7968.85,70
5,0201-OAMXR,115.55,8127.60,70
6,4361-BKAXE,114.50,4527.45,41
7,1555-DJEQW,114.20,7723.90,70
8,9158-VCTQB,113.60,4594.95,41
9,7279-BUYWN,113.20,4689.50,41


In [ ]:
# =====================================================
# FINALIZAÇÃO:
# Fecha a conexão com o banco de dados
# =====================================================

conn.close()

**Perfil de maior risco:**

**Cliente com:**

✔ Contrato mensal

✔ Electronic check

✔ Baixo tenure

✔ Ticket alto

✔ Menor fidelização

**KPIs CRÍTICOS:**

✔ Churn Rate = 26,58%

✔ Receita em risco = $139 mil/mês

✔ Ticket churn maior que ticket ativo

✔ Contratos curtos = maior risco

✔ Electronic check = alto risco

✔ Early churn = forte padrão


# 🧹 6. Limpeza e Tratamento

# 📊 7. Análise Exploratória (EDA) & Insights

# 🤖 8. Modelagem Preditiva (Machine Learning)

# 💡 9. Conclusões e Recomendações de Negócio